### ISSUE TITLE
Introduce DI to Beef

### ISSUE THREAD

**Issue by chullybun** [Avanade/Beef#61]
This issue is to collaborate on making changes to Beef to introduce Dependency Injection (DI) as presented by @edjo23 within his [POC](https://github.com/edjo23/CodeGen.POC). This POC also has suggestions around changing Beef to leverage [Handlebars.NET](https://github.com/rexm/Handlebars.Net) which will managed separately to this change.

This is an initial pass of DI to be used for the generated components only at this stage, being the following:
-  `XxxController` - constructor change for `IXxxManager`.
- `XxxManager` - constructor changes for `IXxxDataSvc`.
- `XxxDataSvc` - constructor changes for `IXxxData`, will no longer be static, and will introduce new `IXxxDataSvc`.

As identified by @edjo23 there can only be a single public constructor, so there needs to be a way to configure this, so the following attributes will be added to the Entity XML:
- `ControllerConstructor="Public|Private|Protected"` - (defaults to `Public`).
- `ManagerConstructor="Public|Private|Protected"` - (defaults to `Public`).
- `DataSvcConstructor="Public|Private|Protected"` - (defaults to `Public`).
- `DataConstructor="Public|Private|Protected"` - (defaults to `Public`).

During discussions @edjo23 suggested a further option to limit the output of the generated extensions code and have the developers explicitly opt-in where required to reduce unnecessary code bloat. This will be all or nothing for the class, not per operation. The following attributes will be added to the Entity XML for those classes that support extensions:
- `ManagerExtensions="true|false"` - defaults to `false`.
- `DataSvcsExtensions="true|false"` - defaults to `false`.
- `DataExtensions="true|false"` - defaults to `false`.

There will be a new `ServiceCollectionExtensions` class generated for (within the appropriate solution hierarchy) to invoke the `IServiceCollection.AddTransient` as follows:
- Api - `./Controllers/ServiceCollectionExtensions.cs` - method `AddGeneratedControllerServices`
- Business - `./Generated/ServiceCollectionExtensions.cs` - method `AddGeneratedManagerServices`
- Business - `./DataSvcs/Generated/ServiceCollectionExtensions.cs` - method `AddGeneratedDataSvcsServices`
- Business - `./Data/Generated/ServiceCollectionExtensions.cs` - method `AddGeneratedDataServices`

Additional changes to be considered:
- The Api changes above need to occur for the gRPC components also.
- The API `startup.cs` will need to be refactored to support DI (including solution templates).
- The testing and mocking will need to be revisited to work with DI.

Please provide feedback and thoughts. A new branch will be created next week to begin work.

Thanks...

**Comment by chullybun** [comment]
Branch [issue-61-introduce-di](https://github.com/Avanade/Beef/tree/issue-61-introduce-di) has been created. A number of the capabilities identified above have been introduced:
- Code-generation DI support for `XxxController`, `XxxManager`, `XxxDataSvc` (instance), `XxxData`, and Grpc `XxxService`.
- Code-generation support for the constructor (per Entity XML config).
- Code-generation support for the extensions (per Entity XML config).
- `IDatabase`, `IEfDb`, `ICosmosDb` and `IOData` interfaces added to enable DI for these into the `XxxData` (removed references to the static `Default` as a result).
- The API `startup.cs` has been updated to enable.

I am super happy with how quickly and easily this has come together so far; and again a huge thanks to @edjo23 for providing the guidance and POC.

Remaining work:
- Add DI support to the `AgentTester` and the internal `TestServer` (where the `IServiceProvider` is managed).
- Review `Agent` and underlying `ServiceAgent` for DI support (considerations for different `HttpClient`, etc.).
- Review the `ReferenceData` implementation to determine changes for DI (firstly assess effort).
- Add DI support for `Functions`; specifically the `EventSubscriber` which is intended to access an `XxxManager`.

Cheers...

**Comment by edjo23** [comment]
Hi @chullybun, I have started looking at the integration tests and how services could be replaced by mocks. You may be way ahead of me but I have updated my POC with an initial concept. It is quite basic and will need some refinement but I thought I'd share.

Thanks

**Comment by chullybun** [comment]
Hey @edjo23, thanks for the updated POC - I was able to follow along and grok the approach. In playing with it on the Beef solution proper the `ReferenceData` with its internal statics just blew up 😢 I think we both expected that. Anyways, this just brought forward the effort to add DI to `ReferenceData`, which again was not as hard as I had expected. To still enable the richness in the entities I had to add the `IServiceProvider` to the `ExecutionContext` - which means I can get access to the configured `ReferenceData` by its interface (new). I looked at the `HttpContext` and they have a similar pattern of keeping a reference to `IServiceProvider` so I don't feel dirty doing this. Will sync once I get the solution building again :-)

**Comment by edjo23** [comment]
Hi @chullybun, I've made another update to the testing stuff in the POC to use ASPNET "options" ([ie](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-3.1)) for injecting parameters into service agents.

Options support the concept of named options allowing for a single type to be used and injected across multiple services. So this allows for one 'ServiceAgentOptions' type to be defined and used across all service agents. The service agents lookup the options based on a name.

I've updated the AgentTesterFactory to use a `ServiceCollection` to configure and build the service agent. So if you are interested in how the options work have a look at the `AgentTesterFactory.Create` method.

HTH

**Comment by edjo23** [comment]
Hi @chullybun, I've made a further change to add custom configuration to the tests.

**Comment by chullybun** [comment]
Hi @edjo23, thanks I had a look. One thing, in using the shared WebApplicationFactory I was expecting it to behave similar to the TestServer we had previously in that it could support multiple concurrent tests (using `[Parallelizable(ParallelScope.Children)]`). But the behaviour is the tests appear to be single-threaded (through the `HttpClient`) as all tests seem to complete in the same time, but not quickly. Are you aware of how to support parallelism?

**Comment by chullybun** [comment]
Hi @edjo23, good progress. Now have all my intra-domain tests passing. Hey one thing that would be super helpful would be to get an `ILogger` that writes to the NUnit `TestContext.Out`. Is this something you could whip up :-) Cheers...

**Comment by edjo23** [comment]
Hi @chullybun, that's awesome. I'll have a look at that logger for you.

With regards to the parallel testing. I had a look, it only appears that the first "batch" of tests have some sort of locking/waiting on a resource, subsequent tests seem to be fine. Still looking into trying to find out what the locking/waiting is to see if it could be primed in the one time set up.

**Comment by edjo23** [comment]
Hi @chullybun, I've created the logger in my POC but it is not working as I hoped. Not sure why but `TestContext.CurrentContext.Test.ID` is different in the test, and in my controller where I have some logging. The result is the logs from the test framework show, but logs from the SUT don't show. Not sure if you wanted to try and log everything, or just stuff directly related to the test (I'm guessing everything).

The code is here if you wanted to try and add it to this repo to see how it goes - https://github.com/edjo23/CodeGen.POC/blob/master/Demo.Service.IntTest/NUnitTestContextLogger.cs

**Comment by chullybun** [comment]
Hi @edjo23. There could be a delay related to the caching of reference data; however, I would expect similar behaviour when using the `TestServer`? I think we just continue supporting `Server` and `Waf` as is. Thanks for the `Logger` I will have a go at integrating and see where I end up (and yes was hoping everywhere). I sync'd up all my changes earlier - all API-related tests are passing - yay! I will start to look at DI in the functions and related event processing. Cheers...

**Comment by chullybun** [comment]
Hi @edjo23. I have the solution running and testing end-to-end. Just the `Beef.Template.Solution` remaining to be fixed up; this is purely cosmetic. I have also updated all the version numbers to `4.0.1` to recognize the significance of the changes. It is in a state now for you to properly review; to make sure we are meeting all the goals we set out in the beginning. As an FYI, the `Factory` class has been removed from the solution as it is no longer being referenced - that is a positive sign!

There are two main objects not instantiated by DI, being `Validators` and `Mappers`. I am in two minds with these, especially the latter. For me they are somewhat internal to the `Manager` and `Data` objects. They are declared as static as these is a runtime cost to instantiate, and they could just as easily be code that would be compiled anyway - the question is whether they would ever need to be mocked? I cannot think of a scenario? Thoughts??

Cheers 🍺

**Comment by edjo23** [comment]
Hi @chullybun, that is fantastic. I'll review it and let you know how I go. I was thinking I might also upgrade some of our components that we have and run their integration tests as well.

With regards to `Validators` and `Mappers` ... I agree with you that they are somewhat internal and I can't think of a scenario either where you would want to mock them.

**Comment by chullybun** [comment]
Hi @edjo23, some more changes made and ready, specifically around testing (improvements/ease of use) resulting in less breaking changes hopefully. Also, as an FYI the `DataContextScope` has been deprecated; connection management now exclusively within `DatabaseBase` (much simpler) - being a connection per instance - which should be scoped per request (the `Startup` uses `AddBeefDatabaseServices` to ensure correct configuration). How is your validating going?

Also, I will probably publish the v4.0.1 packages later this week - you'll need to let folks know not to use yet :-)

**Comment by edjo23** [comment]
Hi @chullybun, I spent some time last night reviewing the changes, got through most of it, so far nothing has jumped out. My plan was to try and get this into some of our existing components so that I can run our existing integration tests. I was going to ask if it would be possible for you to publish pre-release versions of the packages, this would make that much easier for me to do 😄.

**Comment by edjo23** [comment]
Hi @chullybun, I worked out an easy way to deploy pre-release versions of the packages in our stack, so you can ignore that request.

**Comment by edjo23** [comment]
Hi @chullybun, I got one of our smaller solutions cut over to use the new DI changes. It took a lot longer than I thought, mainly due to a bunch of our stuff also needing to be updated. I had planned on getting another larger solutions or two cut over, but ran out of time. Everything still looks good, although I hit a few small issues but managed to work around them.

We have an operation that is a custom, manager only, operation. The XML is defined as:
```xml
<Entity  Name="EventData" ...>
    <Operation  Name="Publish" ... ManagerCustom="true" ExcludeData="true" ExcludeIData="true" ExcludeDataSvc="true" ExcludeIDataSvc="true">
    </Operation>
</Entity>
``` 
... but an IEventDataDataSvc.cs is still created, although not the other files. The ServiceCollectionExtension has this code:

```cs
return services.AddScoped<IEventDataDataSvc, EventDataDataSvc>();
```
... which won't compile as the EventDataDataSvc class has not been generated. The work was easy though, just creating the missing (empty) class.

With regards to testing I couldn't workout how to use the `UsingAgentTesterServer<TStartup>` with a custom configuration, so had to roll my own base class with this setup (since we have a custom configuration setup):
```cs
protected IConfiguration Configuration { get; private set; }
protected AgentTesterServer<Startup> AgentTester { get; private set; }

[OneTimeSetUp]
public void CreateAgentTester()
{
    Configuration = ...;
    AgentTester = Beef.Test.NUnit.AgentTester.CreateServer<Startup>(environment: "IntegrationTest", config: Configuration);
}
```
We also have some tests that make multiple calls via the AgentTester in each test method, e.g. maybe a GET then a PUT, etc. Those tests were failing with this exception:
```
Message: 
    System.ArgumentException : Value is immutable; cannot be changed once already set to a value.
  Stack Trace: 
    ExecutionContext.set_CorrelationId(String value)
    AgentTestCore`1.ctor(AgentTesterBase agentTesterBase, String username, Object args)
    AgentTestBase`1.ctor(AgentTesterBase agentTesterBase, String username, Object args)
    AgentTest`3.ctor(AgentTesterBase agentTesterBase, String username, Object args)
    AgentTesterServer`1.Test[TAgent,TValue](String username, Object args)
```
As a workaround I added `AgentTester.PrepareExecutionContext(username);` between the calls to the AgentTester.

I'll try to get another larger solution updated tomorrow which will have a more comprehensive set of tests, I'll let you know how I go. I also haven't tested any function apps yet (e.g. Event Hub processors), although we largely have our own set of base libraries for those.

Cheers

**Comment by chullybun** [comment]
Hi @edjo23, I've fixed the code-gen where `XxxManager` was expecting `IXxxDataSvc` where none generated. I could not reproduce issue with `ExcludeIDataSvc`; it functioned as expected for me?

I had not made any provisions within `UsingAgentTesterServer` to support custom configuration. Given your usage do you have a suggested implementation, or is it easier for you to have your own customised version? If you offer something similar to `UsingAgentTesterServer` make sure you implement `IUsingAgentTesterServer` as the `TestSetUpAttribute` looks for that and invokes `PrepareExecutionContext` automatically before running the test method. I also have tests with multiple calls to `AgentTester`, these do not require `PrepareExecutionContext`?

I will also publish version 4.0.1-alpha to nuget later today (as I also need for another project I am working on).

Cheers...

**Comment by edjo23** [comment]
Thanks @chullybun, I'll take a look at `IUsingAgentTesterServer` and get back to you.

I'm still trying to get our core multi-tenanted API up and running. It is requiring more effort that I anticipated, there is tons of plumbing code that has been defined as statics in our start up, with lots of new'ed up managers mixed in there for testing auth tokens against tenants and users. So just trying to untangle that.

I hit another couple of smaller issues that I managed to work around, I'll share them tomorrow when I get a chance to write them up.

**Comment by edjo23** [comment]
The issues I had were:

With this config:
```xml
<Entity Name="Tenant" Text="Tenant" ExcludeData="true" ExcludeIData="true" ExcludeDataSvc="true" ExcludeIDataSvc="true" WebApiRoutePrefix="api/v1/tenant">
    ...
</Entity>
<Entity Name="User" Text="User" Implements="IETag, IChangeLog, IUser" Collection="true" CollectionResult="true" WebApiRoutePrefix="api/v1/sec/users" AutoImplement="Database" DatabaseSchema="Sec">
    ...
</Entity>
```

The DataSvc ServiceCollectionExtensions code that is generated doesn't compile.
```cs
        public static IServiceCollection AddGeneratedDataSvcServices(this IServiceCollection services)
        {
                           .AddScoped<IUserDataSvc, UserDataSvc>()
                           .AddScoped<IOrgUnitDataSvc, OrgUnitDataSvc>()
                           .AddScoped<IOrgUnitOpeningHoursDataSvc, OrgUnitOpeningHoursDataSvc>()
                           .AddScoped<IUserProfileDataSvc, UserProfileDataSvc>();
        }
```

I think this is due to the first entity in the code gen xml having `ExcludeIDataSvc="true"`. Swapping the entity definitions around so the the first item does not have `ExcludeIDataSvc="true"` works around the issue.

Another issue was with one (and only one, the last one) of the generated Data access classes injecting in a database parameter of type "Database" not "IDatabase". I didn't track down the root cause but I think the issue is to do with this line - https://github.com/Avanade/Beef/blob/f013fb9c8215d539c0cbdd848b03c3d968eb14c2/tools/Beef.CodeGen.Core/Templates/EntityData_cs.xml#L631

Setting `DatabaseName="IDatabase"` on the root `CodeGeneration` xml element worked around the issue.

Also was wondering if the `GetAllTypes` method generated in the ReferenceData entity class could be made static. We use this list of types to mock a reference data provider for unit testing. Previously we used a new'ed up `ReferenceDataProvider` to get to the property, now with the new constructor for DI this is a bit messier. For now I have just created a class in the test library that inherits the abstract ReferenceData entity class so that I can get to the property, but it would be great not to need to do that.

Cheers
https://github.com/Avanade/Beef/blob/f013fb9c8215d539c0cbdd848b03c3d968eb14c2/samples/Demo/Beef.Demo.Common/Entities/Generated/ReferenceData.cs#L37

**Comment by chullybun** [comment]
Hi @edjo23, I made some changes that I hope will resolve your first two issues - I was unable to replicate behaviour but believe I found the offending code. `GetAllTypes` is now static. I have published new `-alpha-20200805` nuget packages. Let me know how you get on. Cheers...

**Comment by edjo23** [comment]
Hi @chullybun, as it turns out I can't use `IUsingAgentTesterServer` as it is `internal`. I tried to replicate what the attribute does in my base class, using my own attribute for simple metadata and a `[SetUp]` method. This sort of worked but I couldn't call `TestSetUp.InvokeRegisteredSetUp` as it is also `internal`. In the end I had to use reflection to invoke it.

```cs
public abstract class AgentTesterServerTestBase
{
    protected IConfiguration Configuration { get; private set; }
    protected AgentTesterServer<Startup> AgentTester { get; private set; }

    [OneTimeSetUp]
    public void CreateAgentTester()
    {
        TestSetUp.Reset();
        typeof(TestSetUp).GetMethod("InvokeRegisteredSetUp", BindingFlags.Static | BindingFlags.NonPublic).Invoke(null, null);

        Configuration = ...;
        AgentTester = Beef.Test.NUnit.AgentTester.CreateServer<Startup>(environment: "IntegrationTest", config: Configuration);           
        AgentTester.RegisterBeforeRequest(BeforeRequest);
    }

    [SetUp]
    public void Setup()
    {
        var attribute = GetType().GetMethod(TestContext.CurrentContext.Test.MethodName)?.GetCustomAttribute<AgentTesterAttribute>();
        if (attribute != null)
            AgentTester.PrepareExecutionContext(TestSetUp.ConvertUsername(attribute.TestUser), AgentTesterExtensions.GetTenant(attribute.Country));
        else
            AgentTester.PrepareExecutionContext();
    }

    protected virtual void BeforeRequest(HttpRequestMessage request)
    {
        ...
    }
}
```

The next issue I hit was that my `ExecutionContext` was being "lost". I tracked that down to the constructor of the  `AgentTestCore<TStartup>` class.
https://github.com/Avanade/Beef/blob/d418f77abb2d3057da661756108592b7ca5ead54/tools/Beef.Test.NUnit/Tests/AgentTestCore.cs#L38
I think that line should be `if (username != null || !ExecutionContext.HasCurrent)` ... I'm not passing a username as I have an ExecutionContext.

Also I have some other code gen issues with Managers that I'm still looking into, seems to be related to using `ManagerCustom="true"`. 

```xml
    <Operation Name="GetDetails" OperationType="Get" WebApiRoute="details" ReturnType="Tenant" ManagerCustom="true"/>
```
Code is generating as:
```cs
        public Task<Tenant?> GetDetailsAsync()
        {
            return ManagerInvoker.Current.InvokeAsync(this, async () =>
            {
                ExecutionContext.Current.OperationType = OperationType.Read;
                var __result = await GetDetailsOnImplementationAsync().ConfigureAwait(false);
            });
        }
```
I'm having to manually add `return __result;` to get this to compile.

Also:
```xml
    <Operation Name="GetByUsername" OperationType="Custom" ReturnType="User?" WebApiRoute="username/{username}" WebApiMethod="HttpGet" WebApiAlternateStatus="NotFound" AutoImplement="None">
      <Parameter Name="Username" Type="string" />
    </Operation>
```
Is generating as:
```cs
        public Task<User?> GetByUsernameAsync(string? username)
        {
            return ManagerInvoker.Current.InvokeAsync(this, async () =>
            {
                ExecutionContext.Current.OperationType = OperationType.Unspecified;
                Cleaner.CleanUp(username);
                MultiValidator.Create()
                    .Run().ThrowOnError();

                return Cleaner.Clean(await _dataService.GetByUsernameAsync(username).ConfigureAwait(false));
                return Cleaner.Clean(__result);
            });
        }
```
With two return statements.

When I get some time I'll try and create an xml for you to repo the code gen issues.

Cheers 🍺

**Comment by edjo23** [comment]
Hi @chullybun, to reproduce the issue I mentioned above with the code gen'ed DataSvc ServiceCollectionExtensions not compiling; add the following entity as the first entity in the code gen xml.

```xml
  <Entity Name="Tenant" Text="Tenant" ExcludeData="true" ExcludeIData="true" ExcludeDataSvc="true" ExcludeIDataSvc="true" WebApiRoutePrefix="api/v1/tenant">
    <Property Name="Id" Type="Guid" UniqueKey="true" DataName="TenantId"/>
    <Property Name="Name" Type="string" DataName="Name"/>

    <Operation Name="GetDetails" OperationType="Get" WebApiRoute="details" ReturnType="Tenant"  />
  </Entity>
```

**Comment by edjo23** [comment]
Also if you add the xml below (which is slightly different to the previous xml) to the end of the code gen xml, it should repro the other issue I was having.

```xml
  <Entity Name="Tenant" Text="Tenant" ExcludeData="true" ExcludeIData="true" ExcludeDataSvc="true" ExcludeIDataSvc="true" WebApiRoutePrefix="api/v1/tenant">
    <Property Name="Id" Type="Guid" UniqueKey="true" DataName="TenantId"/>
    <Property Name="Name" Type="string" DataName="Name"/>

    <Operation Name="GetDetails" OperationType="Get" WebApiRoute="details" ReturnType="Tenant" ManagerCustom="true"  />
  </Entity>
```
Generated code:
```cs
public Task<Tenant?> GetDetailsAsync()
{
    return ManagerInvoker.Current.InvokeAsync(this, async () =>
    {
        ExecutionContext.Current.OperationType = OperationType.Read;
        var __result = await GetDetailsOnImplementationAsync().ConfigureAwait(false);
    });
}
```

**Comment by chullybun** [comment]
Hi @edjo23, I have renamed `IUsingAgentTesterServer` to `ITestSetupPrepareExecutionContext` and made public, so that this can be used by anyone. I have reviewed where `ExecutionContext.Reset` was being invoked and adjusted where applicable; i.e. should not be being called as often; was perhaps being invoked a little too enthusiastically :-O.

The `Manager` code generation has been reviewed. The first scenario you presented (with no `ManagerCustom="true"`) was working as expected, although `ExcludeDataSvc` and `ExcludeIDataSvc` were selected, as the `Manager` was not custom it still will execute via the interface - it was expecting the developer to have implemented (would therefore exist, just not being automatically generated by Beef). The second scenario with `ManagerCustom="true"` was outputting incorrectly and has been fixed.

Also, not sure if you had referenced, but `Entity.ControllerConstructor` code-gen attribute has been renamed to `Entity.WebApiConstructor` for consistency with other attributes.

I have published new -alpha.20200811 nuget packages.

**Comment by edjo23** [comment]
Thanks for the changes @chullybun.

I'm still having issues with the ExecutionContext username being reset when calling the AgentTesterServer.Test method.

My tests look like this (where `AgentTester` is an `AgentTesterService` instance setup in the class).

```cs
 [Test]
 [AgentTester(TestUser.Admin)]
 public void GetUserSuccessful()
 {
    // Note: ExecutionContext is set up by the SetUp method in the base class (uses the NUnit [SetUp] attribute)

    var user = GetUser();

    AgentTester.Test<UserAgent, User>()
        .IgnoreChangeLog()
        .IgnoreETag()
        .ExpectStatusCode(HttpStatusCode.OK)
        .ExpectValue((t) => user, "Permissions")
        .Run((a) => a.GetAsync(user.Id));
}
```
The parameter docs on the `AgentTesterService.Test` method indicates that passing null for username should use the `ExecutionContext.Current.Username`, but that isn't happening. I think the changes made to the AgentTestCore constructor aren't quite right as they don't handle username being passed as null.

Cheers

**Comment by chullybun** [comment]
Hi @edjo23 - I have attempted to replicate with no luck. I created a couple of [tests](https://github.com/Avanade/Beef/blob/issue-61-introduce-di/samples/Demo/Beef.Demo.Test/AgentTesterTest.cs) to validate and they passed. I assume the `[AgentTester]` attribute is custom as I had renamed in _Beef_ to `[TestSetUp]`. I did another NuGet publish, not that there should be anything in there specifically correcting the issue.

I am planning on doing a `4.1.0` release/publish proper tomorrow, including merge into master. So you will need to let folks know not to use until you are ready.

Cheers...

**Comment by edjo23** [comment]
Hi @chullybun, thanks. I tried to replicate in the Beef.Demo solution, but couldn't. Went back to my stuff and re-checked, and it was working as expected! ... Not sure what happened, I swear I had the debugger on it and it was returning the username as `Anonymous` 😳. Sorry was wasting your time.

The other code gen changes have fixed most of my code gen issues (even one you couldn't reproduce), the only one left is in an area that I'm refactoring anyhow so I think it will go away. If not it is easily worked around.

I've got our integration tests running now and nothing more is jumping out. 

Looking forward to the release 😄.
